File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string b, string e, vector<string>& w) {
        unordered_set<string> s(w.begin(),w.end());
        if(s.find(e)==s.end()) return 0;

        unordered_map<string, vector<string>> nei; // pattern -> vector(words)
        w.push_back(b);
        for(string i: w){
            for(int j=0;j<i.size();j++){
                string pattern=i;
                pattern[j]='*';
                nei[pattern].push_back(i);
            }
        }
        queue<string> q;
        unordered_set<string> vis;
        q.push(b);
        vis.insert(b);
        int ans=1;
        while(!q.empty()){            
            int k=q.size();
            for(int j=0;j<k;j++){
                string word=q.front();
                q.pop();
                if(word==e) return ans;                
                for(int i=0;i<word.size();i++){
                    string pattern=word;
                    pattern[i]='*';
                    for(auto p:nei[pattern]){
                        if(vis.find(p)==vis.end()){
                            vis.insert(p);
                            q.push(p);
                        }
                    }
                }
            }
            ans++;
        }
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0
        nei = collections.defaultdict(list) #adjaceny list .. patter->list(word)
        # collections.defaultdict(list) make sure initial list is empty
        wordList.append(beginWord)

        for word in wordList:
            for j in range(len(word)): 
                #start creating patterns
                pattern = word[:j] + "*" + word[j+1:]
                nei[pattern].append(word)

        vis=set([beginWord])
        q=deque([beginWord])
        ans=1

        while q:
            for i in range(len(q)):
                word=q.popleft()
                if word==endWord:
                    return ans
                for j in range(len(word)): 
                    #start creating patterns
                    pattern = word[:j] + "*" + word[j+1:]
                    for n in nei[pattern]:
                        if n not in vis:
                            vis.add(n)
                            q.append(n)
            ans+=1

        return 0

